File: CS/0797_All_Paths_From_Source_to_Target.java
public class Solution {
    IList<IList<int>> result = new List<IList<int>>();
    bool[] choosed;
    int target;
    int[][] g;
    public IList<IList<int>> AllPathsSourceTarget(int[][] graph) {
        int n = graph.Length;
        choosed = new bool[n];
        target = n-1;
        g = graph;
        
        IList<int> item = new List<int>();
        item.Add(0);
        choosed[0] = true;
        
        backtracking(item, 0);
        
        return result;
    }
    private void backtracking(IList<int> item, int current){
        
        if(current == target){
            result.Add(new List<int>(item));
            return;
        }
        
        int next;
        for(int i = 0; i < g[current].Length; i++){
            next = g[current][i];
            if(choosed[next] == false){
                item.Add(next);
                choosed[next] = true;

                backtracking(item, next);

                item.RemoveAt(item.Count-1);
                choosed[next] = false;
            }
        }
    }
}
/*
public class Solution {
    IList<IList<int>> result = new List<IList<int>>();
    int[][] g;
    int target;
    public IList<IList<int>> AllPathsSourceTarget(int[][] graph) {
        target = graph.Length-1;
        g = graph;
        
        for(int i = 0; i < graph[0].Length; i++){
            IList<int> item = new List<int>();
            item.Add(0);
            helper(item, 0, graph[0][i]);
        }
        
        return result;
    }
    private void helper(IList<int> item, int s, int t){
        item.Add(t);
        if(t == target){
            result.Add(item);
            return;
        }
        else{
            s = t;
            for(int i = 0; i < g[t].Length; i++){
                helper(new List<int>(item), s, g[s][i]);
            }
        }
    }
}
*/

File: Cpp/0797_All_Paths_From_Source_to_Target.cpp
class Solution {
public:
    vector<vector<int>> result;
    int target;
    
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        target = graph.size()-1;
        vector<int> path; 
        path.push_back(0);
        
        helper(result, graph, path, 0);
        return result;
    }
private:
    void helper(vector<vector<int>>& graph, vector<int>& path, int curNode){
        if(curNode == target){
            result.push_back(path);
            return;
        }
        
        for(int i = 0; i < graph[curNode].size(); i++){
            path.push_back(graph[curNode][i]);
            helper(graph, path, graph[curNode][i]);
            path.pop_back();
        }
    }
};
